src/model_repository_manager.cc

```diff
@@ -1938,4 +1990,157 @@ ModelRepositoryManager::DependencyGraph::CircularDependencyCheck(
   return Status::Success;
 }
 
+ModelRepositoryManager::DependencyGraph::DependencyGraph(
+    const DependencyGraph& rhs,
+    std::unordered_map<std::string, std::set<ModelIdentifier>>* global_map)
+    : global_map_ptr_(global_map), missing_nodes_(rhs.missing_nodes_)
+{
+  // Copy nodes_
+  for (auto& pair : rhs.nodes_) {
+    nodes_.emplace(pair.first, std::make_unique<DependencyNode>(*pair.second));
+  }
+  // Re-map pointers from rhs to this
+  for (auto& pair : nodes_) {
+    // upstreams_
+    std::unordered_map<DependencyNode*, std::set<int64_t>> new_upstreams;
+    for (auto& p : pair.second->upstreams_) {
+      new_upstreams.emplace(nodes_.at(p.first->model_id_).get(), p.second);
+    }
+    pair.second->upstreams_.swap(new_upstreams);
+    // downstreams_
+    std::set<DependencyNode*> new_downstreams;
+    for (auto* node : pair.second->downstreams_) {
+      new_downstreams.emplace(nodes_.at(node->model_id_).get());
+    }
+    pair.second->downstreams_.swap(new_downstreams);
+  }
+}
+
+void
+ModelRepositoryManager::DependencyGraph::Assign(
+    const DependencyGraph& rhs,
+    std::unordered_map<std::string, std::set<ModelIdentifier>>* global_map)
+{
+  DependencyGraph tmp(rhs, global_map);
+  Swap(tmp);
+}
+
+void
+ModelRepositoryManager::DependencyGraph::Swap(DependencyGraph& rhs)
+{
+  std::swap(global_map_ptr_, rhs.global_map_ptr_);
```

Comment:
call swap on pointer?

Reply:
This is correct, because the `global_map_` object is owned and managed by the `ModelRepositoryManager` and not by the `DependencyGraph`, so the `DependencyGraph` should only (and always) swap the pointer owned by the `DependencyGraph` (and not the content owned by the `ModelRepositoryManager`). This also avoid a surprising behavior from the `ModelRepositoryManager` standpoint that swapping the `DependencyGraph` causes the content of the `global_map_` object it manages to be updated.